kvserver: sync checksum computation with long poll

pav-kv opened the pull request:
Previously, the checksum computation would run until completion unconditionally
(unless the collection request comes before it). This is not the best spend of
the limited pool capacity, because the result of this computation may never be
requested.

After this commit, the checksum computation task is synchronized with the
checksum collection request. Both wait at most 5 seconds until the other party
has joined. Once joined, the computation starts, otherwise skips.

If any party abandons the request, then the `replicaChecksum` record is preserved
in the state, and is scheduled for a GC later. This is to help the other party
to fail fast, instead of waiting, if it arrives late.

This change also removes the no longer needed concurrency limit for the tasks,
because tasks are canceled reliably and will not pile up.

Fixes #77432

Release note (performance improvement): consistency checks are now properly
cancelled on timeout, preventing them from piling up.

pav-kv committed:
kvserver: sync checksum computation with long poll

Previously, the checksum computation would run until completion unconditionally
(unless the collection request comes before it). This is not the best spend of
the limited pool capacity, because the result of this computation may never be
requested.

After this commit, the checksum computation task is synchronized with the
checksum collection request. Both wait at most 5 seconds until the other party
has joined. Once joined, the computation starts, otherwise skips.

If any party abandons the request, then the replicaChecksum record is preserved
in the state, and is scheduled for a GC later. This is to help the other party
to fail fast, instead of waiting, if it arrives late.

Release note (performance improvement): consistency checks are now properly
cancelled on timeout, preventing them from piling up.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86591)
<!-- Reviewable:end -->

pavelkalinnikov commented:
Issues to address:

- [x] `CollectChecksum` cancel should stop the task. (commit 1)
- [ ] Lack of `CollectChecksum` call in the first place should cause the task to never start (e.g. after a timeout on waiting for it).
- [ ] Clean up / refactor the `replicasChecksum` semantics.

pavelkalinnikov commented:
@tbg @erikgrinaker The first commit partially addresses the issue with minimal code touches. Do you think we could merge or even backport it? Then I could address other things in a follow-up PR.

We can decide after I add more commits to this PR, and it becomes more apparent how we can split it.

pavelkalinnikov commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -466,11 +467,14 @@ func (r *Replica) getChecksum(ctx context.Context, id uuid.UUID) (replicaChecksu
 	if !computed {
-		_, err = r.checksumWait(ctx, id, c.notify, nil)
-		if err != nil {
+		if _, err = r.checksumWait(ctx, id, c.notify, nil); err != nil {
```
:=

pavelkalinnikov commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -70,8 +70,9 @@ var fatalOnStatsMismatch = envutil.EnvOrDefaultBool("COCKROACH_ENFORCE_CONSISTEN
-	started bool
+	// stop terminates the checksum computation. It is not nil iff the computation
+	// has started.
+	stop func()
```
This is created by the computation task, so we can only cancel it from the request if the computation has started. We should also make sure that if the request exited, it marks this `replicaChecksum` in such a way that the computation won't start.

For example, we can store `Context+cancel` here instead, created by whoever first creates this struct. The `Context` will be used by the task, and `cancel` can kill it even before the task has started.

pavelkalinnikov replied:
Added a `gaveUp` field to address this case.

tbg reviewed (changes requested):
I think the main case in which the lack of cancellation caused issues was that caller's context expired before it calls `CollectChecksum`, i.e. the RPC never happens and so the computation at the caller will still be left dangling. So if we merge a partial fix, it should encompass that. The current code doesn't.

I think a decent order of things is

- avoid special-casing of the first checksum (i.e. run all in parallel); hopefully a small change? Let's explore
- add code that aborts the inflight computation unless the long-poll arrives within (say) 5s
- backport the above two
- remove the GC of old checksums, since it's no longer necessary (either long-poll creates it and removes it after computation not starting within 5s, or computation creates it and removes it after long-poll not showing up within 5s, or result of computation gets consumed by long poll and removes it)

This could look something like this:

```go
type bridge struct {
	mu struct {
		syncutil.Mutex
		cancelAll func() // cancels both producer and consumer (if they're there)

		consuming chan struct{} // closed when long poll arrives
		producing chan struct{} // closed when computation starts
		produced  chan struct{} // closed when computation ends (regardless of outcome)
		result    interface{}   // populated with close(produced)
	}
}

func (br *bridge) setProducing() error {
	br.mu.Lock()
	defer br.mu.Unlock()
	select {
	case <-br.mu.producing:
		return errors.AssertionFailedf("already producing")
	default:
	}
	close(br.mu.producing)
	return nil
}

func (br *bridge) setProduced(result interface{}) error {
	br.mu.Lock()
	defer br.mu.Unlock()
	select {
	case <-br.mu.produced:
		return errors.AssertionFailedf("already produced")
	default:
	}
	br.mu.result = result
	close(br.mu.produced)
	return nil
}

func (br *bridge) setConsuming() error {
	br.mu.Lock()
	defer br.mu.Unlock()
	select {
	case <-br.mu.consuming:
		return errors.New("already consuming")
	default:
	}
	close(br.mu.consuming) // NB: mutex makes this non-racy
	return nil
}

func (br *bridge) waitProducerStart(ctx context.Context) error {
	br.mu.Lock()
	ch := br.mu.producing
	br.mu.Unlock()
	select {
	case <-ctx.Done():
		return errors.Wrapf(ctx.Err(), "producer didn't start in time")
	case <-ch:
	}
	return nil
}

func (br *bridge) waitProducerDone(ctx context.Context) (interface{}, error) {
	br.mu.Lock()
	ch := br.mu.produced
	br.mu.Unlock()
	select {
	case <-ctx.Done():
		return nil, errors.Wrapf(ctx.Err(), "while waiting for result")
	case <-ch:
	}
	return br.mu.result, nil // NB: don't need lock here
}

func (br *bridge) isConsuming() bool {
	br.mu.Lock()
	defer br.mu.Unlock()
	select {
	case <-br.mu.consuming:
		return true
	default:
		return false
	}
}

type inflightChecksumMap map[string]*bridge // TODO needs a lock

func (m inflightChecksumMap) getOrCreate(id string, cancelMe func()) *bridge {
	// TODO m.mu.Lock()
	// TODO defer m.mu.Lock()
	if br, ok := m[id]; ok {
		old := br.mu.cancelAll
		br.mu.cancelAll = func() {
			old()
			cancelMe()
		}
		return br
	}
	br := &bridge{}
	br.mu.cancelAll = cancelMe
	br.mu.consuming = make(chan struct{})
	br.mu.producing = make(chan struct{})
	br.mu.produced = make(chan struct{})

	m[id] = br

	return br
}

func (m inflightChecksumMap) delete(id string, ptr *bridge) {
	// TODO m.mu.Lock()
	// TODO defer m.mu.Lock()
	if m[id] != ptr {
		// Shouldn't happen but if for some reason there's an object
		// here that's not the one we worked on, don't do anything
		// as there'll be someone else who's in charge of deleting
		// it and it's simpler not to have to consider this case.
		return
	}
	m[id].mu.cancelAll()
	delete(m, id)
}

var m inflightChecksumMap

func computeChecksum(ctx context.Context, id string) error {
	ctx, cancel := context.WithCancel(ctx)
	br := m.getOrCreate(id, cancel)
	defer m.delete(id, br) // see corresponding line in getChecksum for rationale

	if err := br.setProducing(); err != nil {
		return err
	}

	defer time.AfterFunc(5*time.Second, func() { // TODO use stopper instead
		// NB: if we blocked here, we could delay the computation until the
		// long poll has actually shown up. But we will expect it to usually
		// do that, so probably fine as is? But if it doesn't once then it
		// might happen all of the time. So I would lean towards blocking here.
		if !br.isConsuming() {
			cancel()
		}
	}).Stop()

	var result interface{}
	for i := 0; i < 100; i++ {
		// Big compute! But sensitive to ctx cancellation.
		_ = result
	}
	br.setProduced(result)
}

func getChecksum(ctx context.Context, id string) (interface{}, error) {
	ctx, cancel := context.WithCancel(ctx)
	br := m.getOrCreate(id, cancel)
	// No matter what happens, at the end of this method this ID isn't tracked
	// anymore (though it might be recreated if the computation starts, but then
	// it will time out waiting for getChecksum and will again disappear).
	defer m.delete(id, br)
	if err := br.setConsuming(); err != nil {
		return nil, err
	}
	{
		const startTimeout = 5 * time.Second
		timeoutCtx, cancel := context.WithTimeout(ctx, startTimeout)
		defer cancel()
		if err := br.waitProducerStart(timeoutCtx); err != nil {
			return nil, err
		}
	}
	return br.waitProducerDone(ctx)
}

```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86591#-:-NA9h99zEepI70VF2NlJ:b-5o66yk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86591)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

tbg reviewed (dismissed):
Oops, wrong button.

erikgrinaker reviewed (commented):
>* avoid special-casing of the first checksum (i.e. run all in parallel); hopefully a small change? Let's explore
>* add code that aborts the inflight computation unless the long-poll arrives within (say) 5s
>* backport the above two

It's unclear to me how we can make this safe to backport -- users can presumably run different minor versions in the same cluster, so wouldn't this effectively prevent all consistency checks initiated by older nodes? I was thinking we'd need a major version gate for this, and tolerate slow callers until all nodes have upgraded.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86591#-:-NAE37_m9S2PAyrOGFvS:b66imnz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
Do you mean the following scenario?
- An old node blocks on the first request, so it will typically be more than 5s until other requests arrive.
- A new node expects a timely turnup of a request, so it times out the consistency check.

Possibly we could backport the sequential->parallel change first, coupled with cancel propagation (for the case when the RPC made it through). It's backwards compatible in this regard.

And then we could tackle the timeout behaviour of the task (using the gate etc).

Another backwards-compatible way:
- still don't start the task right away, put it in an in-memory queue.
- allow a longer timeout (same as before).
- let `CollectChecksum` start it when it arrives.
The unclear moment here is: how long is it okay to hold the `snapshot` that the post-apply handler grabs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86591#-:-NAECBobE3cPBPoxSg2J:b1em9jn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
>Possibly we could backport the sequential->parallel change first, coupled with cancel propagation (for the case when the RPC made it through). It's backwards compatible in this regard.

Yeah, as we discussed in the meeting earlier today, we should backport this to 22.2 at least, so that we can rely on the behavior when we add timeouts in 23.1. This would also be a strict improvement, in that we'll at least cancel when there is an RPC, and in the common case there will be an RPC when we run them all in parallel.

pavelkalinnikov commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -441,45 +448,89 @@ func (r *Replica) gcOldChecksumEntriesLocked(now time.Time) {
+	c := r.getReplicaChecksum(ctx, id, now, false)
+	// Schedule GC-ing the record instead of deleting it, to allow a late arriving
+	// checksum computation attempt to see if the caller has given up.
+	defer r.releaseReplicaChecksum(ctx, id, timeutil.Now().Add(replicaChecksumGCInterval))
```
Oups, the `time.Now()` should be evaluated at exit, not here.

pavelkalinnikov commented:
@tbg I factored out backwards-compatible things in #86883, please review it first.

> This could look something like this:

I ended up with just a couple of channels. See: [kvserver: refactor replica checksum comptutation](https://github.com/cockroachdb/cockroach/pull/86591/commits/30b028496ffb656f1c244c2b4a48b575c36cb9de).

pavelkalinnikov commented:
@tbg @erikgrinaker Ready for review now. Disregard all the previous versions of on this PR, that work was factored out and submitted.

erikgrinaker commented on pkg/kv/kvserver/consistency_queue.go:
```diff
@@ -65,22 +65,26 @@ const consistencyCheckRateMinWait = 100 * time.Millisecond
+// consistency check calculations. These are loosely tied to the caller's
+// context, and will be canceled when the caller has given up on them, but we
+// give them an upper timeout to prevent them from running forever anyway.
 const consistencyCheckAsyncTimeout = time.Hour
```
I wonder if we should go ahead and remove this and `consistencyCheckAsyncConcurrency`? They were hacks introduced only because we didn't have proper cancellation, but now that we do (and assuming that it's reliable) I don't think we should need these crutches. `AsyncConcurrency` can also misfire, e.g. if someone is running with RF >= 9.

pavelkalinnikov replied:
I was thinking of the same, because now things don't leak. The only case when build-up is possible is when replicas restart, but this is limited to 5s. So we could either remove the concurrency limit, or make it much higher (just in case many [correlated] failures occur).

erikgrinaker replied:
Let's just remove it outright. A build-up is fine if it's limited to 5 seconds.

pavelkalinnikov replied:
Done.

pavelkalinnikov replied:
Ah wait. I only removed the concurrency limit (and the semaphore) so far (see the second commit). Why is it ok to remove `consistencyCheckAsyncTimeout`? The queue automatically sets a timeout?

pavelkalinnikov replied:
Done both now, see commits 2 and 3.

erikgrinaker commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -51,10 +51,12 @@ import (
-// TODO(pavelkalinnikov): Consider removing GC behaviour in 23.1+, when all the
-// incoming requests are from 22.2+ nodes (hence arrive timely).
+// We still need the delayed GC in order to help a late arriving participant to
+// learn that the other one gave up, and fail fast instead of waiting.
```
If the timeout here is only 5 seconds, I have a slight preference for just getting rid of this. It won't happen in the typical case, and when it does, it's usually an async caller who doesn't really care if it takes 5 seconds to fail.

pavelkalinnikov replied:
Will address this in a follow-up PR, as this one is getting too big :)

pavelkalinnikov replied:
#88902

erikgrinaker reviewed (approved):
Nice! Can we close out #77432 now?

nit: in the release notes, I wouldn't refer to requests or computations, since end users don't know what these internals are. I would reword it something along the lines of "consistency checks are now properly cancelled on timeout, preventing them from piling up."

pavelkalinnikov commented:
> Nice! Can we close out #77432 now?
Marking as fixed by this PR.

> nit: in the release notes, I wouldn't refer to requests or computations, since end users don't know what these internals are. I would reword it something along the lines of "consistency checks are now properly cancelled on timeout, preventing them from piling up."
Done.

pav-kv committed:
kvserver: remove consistency check concurrency limit

Consistency checks require synchronization between the checksum computation
task and the collection requests, the task does not start otherwise. The task
is canceled if the collection request is abandoned.

These factors removed the possibility of an uncontrolled build-up of abandoned
checksum computation tasks competing for resources. All in-flight tasks are
waited on by the requestor.

This commit removes the no longer needed limit for the number of running tasks,
because this is automatically guaranteed by the way consistency checks are run.

Release note: None

pav-kv committed:
kvserver: remove checksum computation task timeout

The consistency queue usually has a much lower timeout (depends on the
consistency checker rate limit etc). This was a fallback for the follower
checks that weren't connected to the queue's context, and could run forever.
Now the task runs iff there is a requestor waiting for its result.

Release note: None

pav-kv committed:
kvserver: unindent the self-termination code block

Release note: None

pavelkalinnikov commented on pkg/kv/kvserver/replica_consistency.go:
```diff
@@ -768,13 +770,26 @@ func (r *Replica) computeChecksumPostApply(
+			log.Errorf(ctx, "checksum collection did not join: %v", err)
+		} else if err := contextutil.RunWithTimeout(ctx, taskName, consistencyCheckAsyncTimeout,
 			func(ctx context.Context) error {
 				defer snap.Close()
```
@erikgrinaker Is there a way to assert that all snapshots were closed in tests?

erikgrinaker replied:
It's possible that #87232 may cover it. Of course, we'll need a test that actually exercises the code path that leaks the snapshot.

@jbowens Do we check for leaked snapshots on engine close, or only leaked iterators?

erikgrinaker replied:
Did a quick test, doesn't look like we do. We should. Opened https://github.com/cockroachdb/pebble/issues/1986.

pavelkalinnikov replied:
> Of course, we'll need a test that actually exercises the code path that leaks the snapshot.

For consistency checks? This bug is fixed, so we could only do it manually (to make sure retrospectively that this bug would have been caught). But there could be tests in Pebble/CRDB for the unhappy case.

erikgrinaker replied:
Yeah, Pebble should check that the detection works.

pavelkalinnikov commented:
bors r=erikgrinaker
